File: app/src/main/java/com/hipoint/snipback/room/repository/AppRepository.java
package com.hipoint.snipback.room.repository;

import android.content.Context;
import android.os.AsyncTask;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModelProviders;

import com.hipoint.snipback.application.AppClass;
import com.hipoint.snipback.room.dao.EventDao;
import com.hipoint.snipback.room.dao.Hd_snipsDao;
import com.hipoint.snipback.room.dao.SnipsDao;
import com.hipoint.snipback.room.db.RoomDB;
import com.hipoint.snipback.room.entities.Event;
import com.hipoint.snipback.room.entities.Hd_snips;
import com.hipoint.snipback.room.entities.Snip;

import java.util.List;

public class AppRepository {
    private EventDao eventDao;
    private LiveData<Event> EventData;
    private Hd_snipsDao hd_snipsDao;
    private LiveData<Hd_snips> HdSnipsData;
    private SnipsDao snipsDao;
    private LiveData<Snip> SnipData;
    private static AppRepository instance;

    public interface OnTaskCompleted{
        void onTaskCompleted(Snip snip);
    }

    public AppRepository(Context context){
//        RoomDB db = AppClass.getAppInsatnce().database;
        RoomDB db = RoomDB.getDatabase(context);
        eventDao = db.eventDao();
        hd_snipsDao=db.hd_snipsDao();
        snipsDao= db.snipsDao();
    }

    public static AppRepository getInstance(){
        if(instance == null) {
            instance = new AppRepository(AppClass.getAppInstance().getContext());
        }
        return instance;
    }


//Event Table Actions START//
    public LiveData<List<Event>> getEventData(){
        return eventDao.getEventData();
    }

    public LiveData<Event> getEventById(int eventId){
        return eventDao.getEventByEventId(eventId);
    }
//data insert
    public void insertEvent(@NonNull Event event){
        new InsertEventAsync(eventDao).execute(event);
    }

    private class InsertEventAsync extends AsyncTask<Event, Void, Void> {
        private EventDao dao;
        public InsertEventAsync(EventDao dao){
            this.dao = dao;
        }

        @Override
        protected Void doInBackground(Event... events) {
            AppClass.getAppInstance().setLastEventId((int) dao.insert(events[0]));
            AppClass.getAppInstance().setLastCreatedEvent(events[0]);
            return null;
        }

    }
//data update
    public void updateEvent(@NonNull Event event){
        new UpdateEventAsync(eventDao).execute(event);
    }

    private class UpdateEventAsync extends AsyncTask<Event, Void, Void> {

        private EventDao dao;
        public UpdateEventAsync(EventDao dao){
            this.dao = dao;
        }

        @Override
        protected Void doInBackground(Event... events) {
            dao.update(events[0]);
            return null;
        }
    }
    //data delete
    public void deleteEvent(@NonNull Event event){
        new DeleteEventAsync(eventDao).execute(event);
    }

    private class DeleteEventAsync extends AsyncTask<Event, Void, Void> {

        private EventDao dao;
        public DeleteEventAsync(EventDao dao){
            this.dao = dao;
        }

        @Override
        protected Void doInBackground(Event... events) {
            dao.delete(events[0]);
            return null;
        }
    }
    public void deleteAllEvent(){
        new DeleteAllEventAsync(eventDao).execute();
    }

    private class DeleteAllEventAsync extends AsyncTask<Void, Void, Void> {

        private EventDao dao;
        public DeleteAllEventAsync(EventDao dao){
            this.dao = dao;
        }

        @Override
        protected Void doInBackground(Void... voids) {
            dao.deleteAll();
            return null;
        }
    }

    //Event Table Actions END//

    //HDSNIP Table Actions START//
    public LiveData<List<Hd_snips>> getHDSnipsData(){
        return hd_snipsDao.getHDSnipsData();
    }

    public void insertHd_snips(@NonNull Hd_snips hd_snips){
        new InsertHDSnipAsync(hd_snipsDao).execute(hd_snips);
    }

    private class InsertHDSnipAsync extends AsyncTask<Hd_snips, Void, Void> {

        private Hd_snipsDao dao;
        public InsertHDSnipAsync(Hd_snipsDao dao){
            this.dao = dao;
        }

        @Override
        protected Void doInBackground(Hd_snips... hd_snips) {
            AppClass.getAppInstance().setLastHDSnipId(dao.insert(hd_snips[0]));
            return null;
        }

    }

    //data update
    public void updateHDSnip(@NonNull Hd_snips hd_snips){
        new UpdateHDSnipAsync(hd_snipsDao).execute(hd_snips);
    }

    private class UpdateHDSnipAsync extends AsyncTask<Hd_snips, Void, Void> {

        private Hd_snipsDao dao;
        public UpdateHDSnipAsync(Hd_snipsDao dao){
            this.dao = dao;
        }

        @Override
        protected Void doInBackground(Hd_snips... hd_snips) {
            dao.update(hd_snips[0]);
            return null;
        }
    }
    //data delete
    public void deleteHDSnip(@NonNull Hd_snips hd_snips){
        new DeleteHDSnipAsync(hd_snipsDao).execute(hd_snips);
    }

    private class DeleteHDSnipAsync extends AsyncTask<Hd_snips, Void, Void> {

        private Hd_snipsDao dao;
        public DeleteHDSnipAsync(Hd_snipsDao dao){
            this.dao = dao;
        }

        @Override
        protected Void doInBackground(Hd_snips... hd_snips) {
            dao.delete(hd_snips[0]);
            return null;
        }
    }
    public void deleteAllHDSnip(){
        new DeleteAllHDSnipAsync(hd_snipsDao).execute();
    }

    private class DeleteAllHDSnipAsync extends AsyncTask<Void, Void, Void> {

        private Hd_snipsDao dao;
        public DeleteAllHDSnipAsync(Hd_snipsDao dao){
            this.dao = dao;
        }

        @Override
        protected Void doInBackground(Void... voids) {
            dao.deleteAll();
            return null;
        }
    }


    //HDSNIP Table Actions END//

    //SNIP table Actions START//
    public LiveData<List<Snip>> getSnipsData(){
        return snipsDao.getSnipsData();
    }

    public void insertSnip(OnTaskCompleted listener,@NonNull Snip snip){
        new InsertSnipAsync(listener,snipsDao).execute(snip);
    }

    private class InsertSnipAsync extends AsyncTask<Snip, Void, Snip> {
        private OnTaskCompleted listener;
        private SnipsDao dao;
        private int snipId;
        public InsertSnipAsync(OnTaskCompleted listener,SnipsDao dao){
            this.listener = listener;
            this.dao = dao;
        }

        @Override
        protected Snip doInBackground(Snip... snips) {
            snipId = (int)dao.insert(snips[0]);
            AppClass.getAppInstance().setLastSnipId(snipId);
            return snips[0];
        }

        @Override
        protected void onPostExecute(Snip aVoid) {
//            aVoid.setSnip_id(AppClass.getAppInstance().getLastSnipId());
//  since lastSnipId is asyncly updated we are not getting the correct snip_id, only the latest one.
            aVoid.setSnip_id(snipId);
            listener.onTaskCompleted(aVoid);
            super.onPostExecute(aVoid);
        }
    }

    //data update
    public void updateSnip(@NonNull Snip snip){
        new UpdateSnipAsync(snipsDao).execute(snip);
    }

    private class UpdateSnipAsync extends AsyncTask<Snip, Void, Void> {

        private SnipsDao dao;
        public UpdateSnipAsync(SnipsDao dao){
            this.dao = dao;
        }

        @Override
        protected Void doInBackground(Snip... snips) {
            dao.update(snips[0]);
            return null;
        }
    }
    //data delete
    public void deleteSnip(@NonNull Snip snip){
        new DeleteSnipAsync(snipsDao).execute(snip);
    }

    private class DeleteSnipAsync extends AsyncTask<Snip, Void, Void> {

        private SnipsDao dao;
        public DeleteSnipAsync(SnipsDao dao){
            this.dao = dao;
        }

        @Override
        protected Void doInBackground(Snip... snips) {
            dao.delete(snips[0]);
            return null;
        }
    }

    public void deleteAllSnip(){
        new DeleteAllSnipAsync(snipsDao).execute();
    }

    private class DeleteAllSnipAsync extends AsyncTask<Void, Void, Void> {

        private SnipsDao dao;
        public DeleteAllSnipAsync(SnipsDao dao){
            this.dao = dao;
        }

        @Override
        protected Void doInBackground(Void... voids) {
            dao.deleteAll();
            return null;
        }
    }
    int eventId = 0;
    public int getLastInsertedEventId(AppCompatActivity activity){
        AppViewModel appViewModel = ViewModelProviders.of(activity).get(AppViewModel.class);
        appViewModel.getEventLiveData().observe(activity, events -> {
            if(events.size() > 0) {
                Event lastEvent = events.get(events.size() - 1);
                eventId = lastEvent.getEvent_id();
                AppClass.getAppInstance().setLastEventId(eventId);
                AppClass.getAppInstance().setLastCreatedEvent(lastEvent);
            }
        });
        return eventId;
    }

    int snipId = 0;
    public int getLastInsertedSnipId(Fragment activity){
        AppViewModel appViewModel = ViewModelProviders.of(activity).get(AppViewModel.class);
        appViewModel.getSnipsLiveData().observe(activity, snips -> {
            if(snips.size() > 0) {
                Snip lastSnip = snips.get(snips.size() - 1);
                snipId = lastSnip.getSnip_id();
                AppClass.getAppInstance().setLastSnipId(snipId);
            }
        });
        return snipId;
    }


    //Snip Table Actions END//
}

File: app/src/main/java/com/hipoint/snipback/room/repository/AppRepository.kt
package com.hipoint.snipback.room.repository

import android.content.Context
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.lifecycle.LiveData
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.hipoint.snipback.application.AppClass
import com.hipoint.snipback.room.dao.EventDao
import com.hipoint.snipback.room.dao.Hd_snipsDao
import com.hipoint.snipback.room.dao.SnipsDao
import com.hipoint.snipback.room.dao.TagDao
import com.hipoint.snipback.room.db.RoomDB
import com.hipoint.snipback.room.entities.Event
import com.hipoint.snipback.room.entities.Hd_snips
import com.hipoint.snipback.room.entities.Snip
import com.hipoint.snipback.room.entities.Tags
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.async
import kotlinx.coroutines.withContext

class AppRepository(context: Context?) {
    private val eventDao: EventDao

    //Event Table Actions START//
    val EventData: LiveData<MutableList<Event>>?
        get() = eventDao.eventData
    private val hd_snipsDao: Hd_snipsDao
    private val HdSnipsData: LiveData<Hd_snips>? = null
    private val snipsDao: SnipsDao
    private val SnipData: LiveData<Snip>? = null
    private val tagDao: TagDao
    private val tagData: LiveData<List<Tags>>
        get() = tagDao.tagsData

    interface OnTaskCompleted {
        suspend fun onTaskCompleted(snip: Snip?)
    }

    interface HDSnipResult{
        suspend fun queryResult(hdSnips: List<Hd_snips>?)
    }

    fun getEventById(eventId: Int): LiveData<Event> {
        return eventDao.getEventByEventId(eventId)
    }

    //data insert
    suspend fun insertEvent(event: Event) {
        withContext(IO) {
            AppClass.getAppInstance().lastEventId = eventDao.insert(event).toInt()
            AppClass.getAppInstance().setLastCreatedEvent(event)
        }
    }

    //data update
    suspend fun updateEvent(event: Event) {
//        UpdateEventAsync(eventDao).execute(event)
        withContext(IO) {
            eventDao.update(event)
        }
    }

    //data delete
    suspend fun deleteEvent(event: Event) {
//        DeleteEventAsync(eventDao).execute(event)
        withContext(IO) {
            eventDao.delete(event)
        }
    }

    suspend fun deleteAllEvent() {
//        DeleteAllEventAsync(eventDao).execute()
        withContext(IO) { eventDao.deleteAll() }
    }

    //Event Table Actions END//
    //HDSNIP Table Actions START//
    val hDSnipsData: LiveData<List<Hd_snips>>
        get() = hd_snipsDao.hdSnipsData

    suspend fun insertHd_snips(hd_snips: Hd_snips) {
//        InsertHDSnipAsync(hd_snipsDao).execute(hd_snips)
        withContext(IO) {
            AppClass.getAppInstance().lastHDSnipId = hd_snipsDao.insert(hd_snips)
        }
    }

    //data update
    suspend fun updateHDSnip(hd_snips: Hd_snips) {
//        UpdateHDSnipAsync(hd_snipsDao).execute(hd_snips)
        withContext(IO) {
            hd_snipsDao.update(hd_snips)
        }
    }

    //data delete
    suspend fun deleteHDSnip(hd_snips: Hd_snips) {
//        DeleteHDSnipAsync(hd_snipsDao).execute(hd_snips)
        withContext(IO) {
            hd_snipsDao.delete(hd_snips)
        }
    }

    suspend fun deleteAllHDSnip() {
//        DeleteAllHDSnipAsync(hd_snipsDao).execute()
        withContext(IO) {
            hd_snipsDao.deleteAll()
        }
    }

    suspend fun getHDSnipsBySnipID(listener: HDSnipResult, snipId: Int){
        val result = CoroutineScope(IO).async{
            hd_snipsDao.getBySnipId(snipId)
        }
        listener.queryResult(result.await())
    }

    suspend fun getHDSnipById(listener: HDSnipResult, hdSnipId: Int){
        val result = CoroutineScope(IO).async{
            hd_snipsDao.getById(hdSnipId)
        }
        listener.queryResult(mutableListOf(result.await()))
    }

    suspend fun getAllHDSnips(): MutableList<Hd_snips>? {
        val result = CoroutineScope(IO).async {
            hd_snipsDao.getAll()
        }
        return result.await()
    }

    //HDSNIP Table Actions END//
    //SNIP table Actions START//
    val snipsData: LiveData<List<Snip>>
        get() = snipsDao.snipsData

    suspend fun insertSnip(listener: OnTaskCompleted, snip: Snip) {
//        InsertSnipAsync(listener, snipsDao).execute(snip)
        val result = CoroutineScope(IO).async {
            val appInstance = AppClass.getAppInstance()
            appInstance.lastSnipId = snipsDao.insert(snip).toInt()
            snip.snip_id = AppClass.getAppInstance().lastSnipId
            appInstance.snips.add(snip)
            snip
        }

        listener.onTaskCompleted(result.await())
    }

    suspend fun getSnipById(snipId: Int): Snip {
        val snip = CoroutineScope(IO).async {
            snipsDao.getSnipById(snipId)
        }
        return snip.await()
    }

    suspend fun getSnipByVideoPath(videoPath: String): Snip {
        val snip = CoroutineScope(IO).async {
            snipsDao.getSnipByVideoPath(videoPath)
        }
        return snip.await()
    }

    //data update
    suspend fun updateSnip(snip: Snip) {
//        UpdateSnipAsync(snipsDao).execute(snip)
        withContext(IO) {
            snipsDao.update(snip)
        }
    }

    //data delete
    suspend fun deleteSnip(snip: Snip) {
//        DeleteSnipAsync(snipsDao).execute(snip)
        withContext(IO) {
            snipsDao.delete(snip)
        }
    }

    suspend fun deleteAllSnip() {
//        DeleteAllSnipAsync(snipsDao).execute()
        withContext(IO) {
            snipsDao.deleteAll()
        }
    }

    suspend fun getAllSnips(): MutableList<Snip>? {
        val result = CoroutineScope(IO).async {
            snipsDao.getAll()
        }
        return result.await()
    }

    var eventId = 0
    fun getLastInsertedEventId(activity: AppCompatActivity?): Int {
        val appViewModel = ViewModelProvider(activity!!).get(AppViewModel::class.java)
        appViewModel.eventLiveData.observe(activity, Observer { events: List<Event> ->
            if (events.isNotEmpty()) {
                val lastEvent = events[events.size - 1]
                eventId = lastEvent.event_id
                AppClass.getAppInstance().lastEventId = eventId
                AppClass.getAppInstance().setLastCreatedEvent(lastEvent)
            }
        })
        return eventId
    }

    var snipId = 0
    fun getLastInsertedSnipId(activity: Fragment?): Int {
        val appViewModel = ViewModelProvider(activity!!).get(AppViewModel::class.java)
        appViewModel.snipsLiveData.observe(activity, Observer { snips: List<Snip> ->
            if (snips.isNotEmpty()) {
                val lastSnip = snips[snips.size - 1]
                snipId = lastSnip.snip_id
                AppClass.getAppInstance().lastSnipId = snipId
            }
        })
        return snipId
    } //Snip Table Actions END//


    //  Tags data
    suspend fun insertTag(tag: Tags) {
//        InsertSnipAsync(listener, snipsDao).execute(snip)
        val result = CoroutineScope(IO).async {
            tagDao.insert(tag)
        }
    }

    suspend fun getTagById(tagId: Int): Tags? {
        val tag = CoroutineScope(IO).async {
            tagDao.getTagById(tagId)
        }
        return tag.await()
    }

    suspend fun getTagBySnipId(snipId: Int): Tags? {
        val tag = CoroutineScope(IO).async {
            tagDao.getTagBySnipId(snipId)
        }
        return tag.await()
    }

    //data update
    suspend fun updateTag(tag: Tags) {
        withContext(IO) {
            tagDao.update(tag)
        }
    }

    //data delete
    suspend fun deleteTag(tag: Tags) {
        withContext(IO) {
            tagDao.delete(tag)
        }
    }

    suspend fun deleteAllTags() {
        withContext(IO) {
            tagDao.deleteAll()
        }
    }

    suspend fun getAllTags(): MutableList<Tags>? {
        val result = CoroutineScope(IO).async {
            tagDao.getAll()
        }
        return result.await() as? MutableList<Tags>
    }

    suspend fun getSnipIdsByColour(colourTagName: String): List<Int>?{
        val result = CoroutineScope(IO).async {
            tagDao.getSnipIdsByColour(colourTagName)
        }

        return result.await()
    }

    suspend fun getSnipIdsByShareLater(): List<Int>?{
        val result = CoroutineScope(IO).async {
            tagDao.getSnipIdsByShareLater()
        }

        return result.await()
    }

    suspend fun getSnipIdsByLinkLater(): List<Int>?{
        val result = CoroutineScope(IO).async {
            tagDao.getSnipIdsByLinkLater()
        }

        return result.await()
    }

    suspend fun getSnipIdsByTextTag(tagText: String): List<Int>?{
        val result = CoroutineScope(IO).async {
            tagDao.getSnipIdsByTextTag(tagText)
        }

        return result.await()
    }


    companion object {
        @JvmStatic
        var instance: AppRepository? = null
            get() {
                if (field == null) {
                    field = AppRepository(AppClass.getAppInstance().context)
                }
                return field
            }
            private set
    }

    init {
//        RoomDB db = AppClass.getAppInsatnce().database;
        val db      = RoomDB.getDatabase(context)
        eventDao    = db.eventDao()
        hd_snipsDao = db.hd_snipsDao()
        snipsDao    = db.snipsDao()
        tagDao      = db.tagDao()
    }
}